### PR TITLE
Adjust http client to work old url format like a[]=1&a[]=2

### PR DESCRIPTION
**Pull request by ecrisac**
In previous version it was not possible to work with old fashion params like a[]=1&a[]=2 in was capturing  just last  parameter in parsed url object

### CODE REVIEW

File: src/url.ts [hunk 1]
```diff
@@ -237,8 +247,10 @@ export namespace Url {
       .map(part => part.split("="))
       .reduce(
         (memo, [param, value]) => {
+          const decodedName = decode(param);
+          const decodedValue = decode(value);
           if (param) {
-            memo[decode(param)] = decode(value);
+            memo[decodedName] ? memo[decodedName].push(decodedValue) : (memo[decodedName] = [decodedValue]);
```
Comment (ardevelop) [inline src/url.ts]: I'd prefer such variant (keep consistent variable names `param` ↔ `name`, and make it a bit easier to read):

```ts
(params, [encodedName, encodedValue]) => {
  const name = decode(encodedName);
  const value = decode(encodedValue);
  const values = memo[name] || (memo[name] = []);

  values.push(value);

  return params;
}
```

I'd drop `if (param)` to be closer with built in `URL` class:

```js
(new URL("http://example.com?=foo")).searchParams.getAll("");
// returns ["foo"]
```

File: src/url.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
-export type TUrlRawQuery = Record<string, string>;
+export type TUrlRawQuery = Record<string, string[]>;
 
-export type TUrlQuery = Record<string, IStringifyable | undefined>;
+export type TUrlQuery = Record<string, IStringifyable | string[] | undefined>;
```
Comment (ardevelop) [inline src/url.ts]: I'd introduce/rename one type to match the property name it describe

```ts
type TUrlParams = Record<string, string[]>;
```

File: src/url.ts [hunk 3]
```diff
@@ -118,24 +118,34 @@ export class Url {
...
+          memo[key] ? memo[key].push(value.toString()) : (memo[key] = [value.toString()]);
+        }
 
-      return memo;
-    }, {});
+        return memo;
+      },
+      {} as Record<string, string[]>
+    );
   }
```
Comment (ardevelop) [inline src/url.ts:127]: Why not just filter params with empty array value ?

```ts
class Url {
  // ...

  get nonNullParams(): TUrlParams {
    const {params = {}} = this.parts;

    return Object.keys(params).reduce((result: TUrlParams , name: string) => {
      const value = params[name];

      if (value && value.length) {
        result[name] = value;
      }

      return result;
    }, {} as TUrlParams);
  }

  // ...
}
```

File: src/url.ts [hunk 4]
```diff
@@ -118,24 +118,34 @@ export class Url {
...
     return Object.keys(this.params || {})
       .sort((a, b) => (a < b ? -1 : a === b ? 0 : 1))
       .reduce((memo: string[], key: string) => {
-        if (this.params[key] != null) {
-          memo.push(`${key}=${encodeURIComponent(this.params[key]!.toString())}`);
+        const value = this.params[key];
+
+        if (value != null) {
+          const values = value instanceof Array ? value : [value];
+          values.forEach(valueItem => memo.push(`${key}=${encodeURIComponent(valueItem.toString())}`));
```
Comment (ardevelop) [inline src/url.ts]: Most probably we should encode `key` as well

File: src/url.ts (hunk 3 above)
Comment (ardevelop) [reply src/url.ts:127]: Not sure this is critical functionality for this class to return filtered params, so most probably we can drop this property as we already do a major change in this library

File: src/url.ts [hunk 5]
```diff
@@ -118,24 +118,34 @@ export class Url {
...
+    );
   }
 
   get paramsAsString(): string {
     return Object.keys(this.params || {})
       .sort((a, b) => (a < b ? -1 : a === b ? 0 : 1))
       .reduce((memo: string[], key: string) => {
-        if (this.params[key] != null) {
-          memo.push(`${key}=${encodeURIComponent(this.params[key]!.toString())}`);
+        const value = this.params[key];
```
Comment (ardevelop) [inline src/url.ts]: - Looks like we're cloning `this.parts.params` each time in loop.

File: src/url.ts [hunk 6]
```diff
@@ -118,24 +118,34 @@ export class Url {
...
+
+        if (value != null) {
```
Comment (ardevelop) [inline src/url.ts]: Should types guarantee that is an array here ?

File: src/url.ts [hunk 7]
```diff
@@ -118,24 +118,34 @@ export class Url {
...
+          const values = value instanceof Array ? value : [value];
+          values.forEach(valueItem => memo.push(`${key}=${encodeURIComponent(valueItem.toString())}`));
         } else {
-          memo.push(`${key}`);
+          memo.push(key);
```
Comment (ardevelop) [inline src/url.ts]: I guess same here, `key` should be encoded, I guess we need to do it only in one line

File: src/url.ts (hunk 3 above)
Comment (procopenco) [reply src/url.ts:127]: it will not work for `?color[]=white&color[]=black` 
we can have multiple params with same name

Comment (ardevelop) [reply src/url.ts:127]: Below It's how I'm expecting it to work: did I miss sth ?

```ts
const regular = Url.fromString("https://example.com?color[]=white&color[]=black");

regular.params // {"color[]": ["white", "black"]}

const withEmptyParam = Url.fromString("https://example.com?color[]=white&color[]=black&color[]=");

withEmptyParam.params // {"color[]": ["white", "black", ""]}
```

File: src/url.ts [hunk 8]
```diff
@@ -1,47 +1,55 @@
...
+    const keys = Object.keys(oldParams).concat(Object.keys(params));
+
+    const newParams = keys.reduce(
+      (memo: TUrlParams, name: string) => {
+        !memo[name] && (memo[name] = (oldParams[name] || []).concat(params[name] || []));
+
+        return memo;
+      },
+      {} as TUrlParams
+    );
```
Comment (ardevelop) [inline src/url.ts:46]: What do you think about this ?

```ts
public appendParams(nextParams: TUrlParams): Url {
  const {params: prevParams = {}} = this._parts;

  const params = Object.keys(nextParams).reduce(
    (memo, name) => {
      memo[name] = (memo[name] || []).concat(nextParams[name]);

      return memo;
    },
    {...prevParams}
  );

  return new Url({...this._parts, params});
}
```

**Review by ardevelop — APPROVED** [review]

**Pull request closed without merging**